test/libyul/ControlFlowGraphTest.cpp

```diff
@@ -0,0 +1,270 @@
+				m_stream << "Jump\" shape=oval];\n";
+				m_stream << "Block" << getBlockId(_block) << "Exit -> Block" << getBlockId(*_jump.target) << ";\n";
+			},
+			[&](CFG::BasicBlock::ConditionalJump const& _conditionalJump)
+			{
+				m_stream << "Block" << getBlockId(_block) << " -> Block" << getBlockId(_block) << "Exit;\n";
+				m_stream << "Block" << getBlockId(_block) << "Exit [label=\"{ ";
+				m_stream << stackSlotToString(_conditionalJump.condition);
+				m_stream << "| { <0> Zero | <1> NonZero }}\" shape=Mrecord];\n";
+				m_stream << "Block" << getBlockId(_block);
+				m_stream << "Exit:0 -> Block" << getBlockId(*_conditionalJump.zero) << ";\n";
+				m_stream << "Block" << getBlockId(_block);
+				m_stream << "Exit:1 -> Block" << getBlockId(*_conditionalJump.nonZero) << ";\n";
+			},
+			[&](CFG::BasicBlock::FunctionReturn const& _return)
+			{
+				m_stream << "Block" << getBlockId(_block) << "Exit [label=\"FunctionReturn[" << _return.info->function.name.str() << "]\"];\n";
+				m_stream << "Block" << getBlockId(_block) << " -> Block" << getBlockId(_block) << "Exit;\n";
+			},
+			[&](CFG::BasicBlock::Terminated const&)
+			{
+				m_stream << "Block" << getBlockId(_block) << "Exit [label=\"Terminated\"];\n";
+				m_stream << "Block" << getBlockId(_block) << " -> Block" << getBlockId(_block) << "Exit;\n";
+			}
+		}, _block.exit);
+		m_stream << "\n";
+	}
+	size_t getBlockId(CFG::BasicBlock const& _block)
+	{
+		if (size_t* id = util::valueOrNullptr(m_blockIds, &_block))
+			return *id;
+		size_t id = m_blockIds[&_block] = m_blockCount++;
+		m_blocksToPrint.emplace_back(&_block);
+		return id;
+	}
+	std::ostream& m_stream;
+	std::map<CFG::BasicBlock const*, size_t> m_blockIds;
+	size_t m_blockCount = 0;
+	std::list<CFG::BasicBlock const*> m_blocksToPrint;
+};
+
+TestCase::TestResult ControlFlowGraphTest::run(ostream& _stream, string const& _linePrefix, bool const _formatted)
+{
+	ErrorList errors;
+	auto [object, analysisInfo] = parse(m_source, *m_dialect, errors);
+	if (!object || !analysisInfo || !Error::containsOnlyWarnings(errors))
+	{
+		AnsiColorized(_stream, _formatted, {formatting::BOLD, formatting::RED}) << _linePrefix << "Error parsing source." << endl;
+		printErrors(errors);
+		return TestResult::FatalError;
+	}
+
+	std::ostringstream output;
+
+	std::unique_ptr<CFG> cfg = ControlFlowGraphBuilder::build(*analysisInfo, *m_dialect, *object->code);
+
+	output << "digraph CFG {\nnodesep=0.7;\nnode[shape=box];\n\n";
+	ControlFlowGraphPrinter printer{output};
+	printer(*cfg->entry);
+	for (auto function: cfg->functions)
+		printer(cfg->functionInfo.at(function));
+	output << "}\n";
+
+	m_obtainedResult = output.str();
+
+	auto result = checkResult(_stream, _linePrefix, _formatted);
+
+#ifdef ISOLTEST
+	char* graphDisplayer = nullptr;
+	if (result == TestResult::Failure)
+		graphDisplayer = getenv("ISOLTEST_DISPLAY_GRAPHS_FAILURE");
+	else if (result == TestResult::Success)
+		graphDisplayer = getenv("ISOLTEST_DISPLAY_GRAPHS_SUCCESS");
```

Comment:
I know it's already long, but having `COMMAND` in the name makes it clearer what the expected value of the variable is (with the current name I would not really expect it to be the name of an executable):
```suggestion
	if (result == TestResult::Failure)
		graphDisplayer = getenv("ISOLTEST_FAILURE_DISPLAY_GRAPHS_COMMAND");
	else if (result == TestResult::Success)
		graphDisplayer = getenv("ISOLTEST_SUCCESS_DISPLAY_GRAPHS_COMMAND");
```
- We should document these variables somewhere.
- Why different commands for success and failure?
- The program you use for this is `dot`, right? Maybe it should be the default value?

Reply:
Displaying something in the success and failure case have rather orthogonal use cases.
If you work on this stuff and change something, you'll want to have it show you the failure cases and just pass through on the success cases.
If you review a PR and want to verify test cases, you'll want to have it stop at each case and show you the graph.
I don't think there's a reasonable default value for this.
You can use ``dot -Tx11:cairo``, but that's already platform specific and sucks interface wise. You can use ``xdot``, if you have it installed and it's available on your platform, but you know neither.

I don't think this will actually be done regularly later on, so I don't think it's worth over-engineering.

Reply:
Renamed them in any case.

Reply:
Also added some comments.